File: cpp/src/advanced/template_class.cpp
#include "template_class.h"



template <typename T>
void Container<T>::set_value(T v) { value = v; }
template <typename T>
T Container<T>::get_value() const { return value; }

void template_example(){

  Container<int> intContainer;
  intContainer.set_value(42);
  cout << "Int value: " << intContainer.get_value() << endl;

  Container<float> floatContainer;
  floatContainer.set_value(3.14f);
  cout << "Float value: " << floatContainer.get_value() << endl;

  Container<string> stringContainer;
  stringContainer.set_value("Hello, World!");
  cout << "String value: " << stringContainer.get_value() << endl;
}

File: rust/src/advanced/template_class.rs

// come una Union
struct Container<T> { value: T, }

impl<T> Container<T> {
  fn new(value: T) -> Self { Container { value }  }
  fn set_value(&mut self, value: T) { self.value = value;  }
  fn get_value(&self) -> &T { &self.value }
}

pub fn template_example() {

  let mut p = Container::new(0);
  p.set_value(42);
  println!("Int value: {}", p.get_value());

  let p = Container::new(3.14);
  println!("Float value: {}", p.get_value());

  let p = Container::new("Hello World!");
  println!("String value: {}", p.get_value());
}
